GRC: generate valid C++ string parameters without explicit double quotes

AsriFox opened the pull request:
## Description
When generating C++ code, GRC did not have a standard way to translate Python strings to C++ strings, because Mako uses simple text substitution. To address this, additional translations are added to generators for blocks and parameters/variables with the first commit.

Translations already used in the block definitions were removed with the second commit:
- manual translations with `no_quotes` helper, like in [qtgui::waterfall_sink_x](https://github.com/gnuradio/gnuradio/blob/9b3183aec7b629fd7fbb2b1574f0bb1c87eb53ee/gr-qtgui/grc/qtgui_waterfall_sink_x.block.yml#L329);
- regex translations in `cpp_templates`, like in [audio::source](https://github.com/gnuradio/gnuradio/blob/9b3183aec7b629fd7fbb2b1574f0bb1c87eb53ee/gr-audio/grc/audio_source.block.yml#L48):
```
translations:
  "'": '"'
  True: true
  False: false
```

## Related Issue
Fixes #2519
Fixes #6551

## Which blocks/areas does this affect?
GRC
Block definition files in: gr-analog, gr-audio, gr-blocks, gr-digital, gr-fft, gr-filter, gr-network, gr-qtgui, gr-zeromq

## Testing Done
Generated C++ code from this flowgraph:
[cpp_string_bug_fixed.grc.txt](https://github.com/gnuradio/gnuradio/files/13464281/cpp_string_bug_fixed.grc.txt)

Issue fixed for simple and widely-used cases:
| Condition | GRC | Display | C++ Output |
|-|-|-|-|
| String parameter is empty | `value: ''` | | `""` |
| String parameter is set to empty single-quotes | `value: ''''''` | | `""` |
| String parameter is set to empty double-quotes | `value: '""'` | | `""` |
| String parameter with no quotes | `value: 127.0.0.1` | 127.0.0.1 | `"127.0.0.1"` |
| String parameter in single quotes | `value: '''127.0.0.1'''` | 127.0.0.1 | `"127.0.0.1"` |
| String parameter in double quotes | `value: '"127.0.0.1"'` | 127.0.0.1 | `"127.0.0.1"` |
| Multi-part string in double quotes | `value: '"127.0"".0.1"'` | 127.0.0.1 | `"127.0"".0.1"` |
| Escaped double quotes in quoted strings | `value: '''127.0.\"\"0.1'''` | 127.0"".0.1 | `"127.0\"\".0.1"` |

Problems arise for these cases, because substitution is regex-based:
| Condition | GRC | Display | C++ Output |
|-|-|-|-|
| Multi-part string in single quotes | `value: '''127.0.''''0.1'''` | 127.0.0.1 | `"127.0''.0.1"` |
| Escaped single quotes in quoted strings* | `value: '''127.0.\''\''0.1'''` | 127.0''.0.1 | `"127.0\"\".0.1"` |
| Raw string literals** | `value: R'127.0.0.1'` | 127.0.0.1 | `R"127.0.0.1"` |

\* Escaped single-quotes are translated with regex when used as values for block parameters. Variable/parameter blocks do not use the regex, so they produce the correct result: "127.0.\'\'0.1".

\** C++ raw string literals require delimiters (parenthesis, etc.), while Python raw string literals don't. This C++ output does not compile.

## Checklist
- [x] I have read the [CONTRIBUTING document](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md).
- [x] I have squashed my commits to have one significant change per commit. 
- [x] I [have signed my commits before making this PR](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md#dco-signed)
- [x] My code follows the code style of this project. See [GREP1.md](https://github.com/gnuradio/greps/blob/main/grep-0001-coding-guidelines.md).
- [ ] I have updated [the documentation](https://wiki.gnuradio.org/index.php/Main_Page#Documentation) where necessary.
- [x] I have added tests to cover my changes, and all previous tests pass.

willcode commented:
Check `tools/autoformat_python.sh` and `tools/check_pep8.sh` for Python formatting rules.

If the rules have strayed from the one use by CI, the current args should be
`--max-line-length=120 --ignore E265,E266,E275,E402,E501,E704,E712,E713,E714,E711,E721,E722,E741,W504,W605`

AsriFox commented:
GRC test was failing because of a check I added in [Block.format_expr](https://github.com/AsriFox/gnuradio/blob/614b3633c43fba79cffa494e4a5249d9e33c7524/grc/core/blocks/block.py#L457). It is now fixed.
However, `int` and `long` type distinction is specific to C++, and #6964 removes it. Maybe I should remove it as well, and simply use `long` for all integers?

willcode commented on grc/core/blocks/block.py:
```diff
@@ -451,71 +449,47 @@ def get_type(element, _vtype):
             self.cpp_templates['includes'].append('#include <string>')
 
-    def get_cpp_value(self, pyval, vtype: str):
+    def get_cpp_value(self, pyval, _vtype: typing.Optional[type] = None):
```
Why the `_` in the parameter name?

willcode commented on grc/core/blocks/block.py:
```diff
@@ -442,49 +440,35 @@ def format_expr(self, py_type):
 
         # Determine the lvalue type
-        def get_type(element, _vtype):
+        def get_type(element: str, _vtype: typing.Optional[type] = None) -> str:
```
Why `_` in `_vtype`?

AsriFox replied:
Left it as it was before me ([source](https://github.com/gnuradio/gnuradio/blob/d7f887227add25431b948d3a18471efd56bb1945/grc/core/blocks/block.py#L443)). Perhaps, it denoted an optional argument?

haakov commented:
Thank you for the PR! I see that you've marked this as a draft, are there other things you want to add before removing the draft tag?

AsriFox commented:
I am unsure what to do with irregular strings (multi-part, escaped quotes, raw literals).
When using my example, in generated C++ code there are compilation errors in cases 8 (escaped single-quotes) and 9 (escaped double-quotes); in generated Python code, `zeromq_req_source_9` throws an `Invalid argument` error (escaped double-quotes).

willcode commented:
Should this be set back to "draft", or have the items above been addressed?

AsriFox commented:
Removed underscores.

haakov commented:
I see. Is there a reason why we should expect those irregular strings in flowgraphs? I'm not sure I've seen one.

If not, I'm inclined to merge this as-is. (Once the conflict with grc/core/blocks/block.py has been resolved, that is)

AsriFox commented:
I expect those irregular strings to be rare. Raw literals and escaping characters might be handled later.

haakov reviewed (approved):
Looks good! But these commits should be squashed before merging.
